File: backend/app/api/tasks.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from datetime import datetime

from app.db.database import get_db
from app.db.models import TaskListModel, TaskModel
from app.schemas.task import (
    Task,
    TaskCreate,
    TaskList,
    TaskListCreate,
    TaskListUpdate,
    TaskNotesUpdate,
    TaskUpdate,
)
router = APIRouter(tags=["tasks"])


@router.get("/task-lists", response_model=list[TaskList])
def get_task_lists(db: Session = Depends(get_db)):
    return db.query(TaskListModel).order_by(TaskListModel.created_at.desc()).all()


@router.post("/task-lists", response_model=TaskList)
def create_task_list(
    task_list_data: TaskListCreate,
    db: Session = Depends(get_db),
):
    if not task_list_data.name.strip():
        raise HTTPException(status_code=400, detail="Task list name is required")

    task_list = TaskListModel(
        name=task_list_data.name.strip(),
        description=task_list_data.description.strip()
        if task_list_data.description
        else None,
    )

    db.add(task_list)
    db.commit()
    db.refresh(task_list)

    return task_list

@router.patch("/task-lists/{task_list_id}", response_model=TaskList)
def update_task_list(
    task_list_id: int,
    task_list_data: TaskListUpdate,
    db: Session = Depends(get_db),
):
    task_list = (
        db.query(TaskListModel)
        .filter(TaskListModel.id == task_list_id)
        .first()
    )

    if task_list is None:
        raise HTTPException(status_code=404, detail="Task list not found")

    if not task_list_data.name.strip():
        raise HTTPException(status_code=400, detail="Task list name is required")

    task_list.name = task_list_data.name.strip()
    task_list.description = (
        task_list_data.description.strip()
        if task_list_data.description
        else None
    )

    db.commit()
    db.refresh(task_list)

    return task_list


@router.delete("/task-lists/{task_list_id}")
def delete_task_list(task_list_id: int, db: Session = Depends(get_db)):
    task_list = (
        db.query(TaskListModel)
        .filter(TaskListModel.id == task_list_id)
        .first()
    )

    if task_list is None:
        raise HTTPException(status_code=404, detail="Task list not found")

    db.delete(task_list)
    db.commit()

    return {
        "status": "ok",
        "message": "Task list deleted successfully",
    }
    
@router.get("/task-lists/{task_list_id}/tasks", response_model=list[Task])
def get_tasks(task_list_id: int, db: Session = Depends(get_db)):
    task_list = (
        db.query(TaskListModel)
        .filter(TaskListModel.id == task_list_id)
        .first()
    )

    if task_list is None:
        raise HTTPException(status_code=404, detail="Task list not found")

    return (
        db.query(TaskModel)
        .filter(TaskModel.task_list_id == task_list_id)
        .order_by(TaskModel.created_at.desc())
        .all()
    )


@router.post("/task-lists/{task_list_id}/tasks", response_model=Task)
def create_task(
    task_list_id: int,
    task_data: TaskCreate,
    db: Session = Depends(get_db),
):
    task_list = (
        db.query(TaskListModel)
        .filter(TaskListModel.id == task_list_id)
        .first()
    )

    if task_list is None:
        raise HTTPException(status_code=404, detail="Task list not found")

    if not task_data.title.strip():
        raise HTTPException(status_code=400, detail="Task title is required")

    task = TaskModel(
        task_list_id=task_list_id,
        title=task_data.title.strip(),
        is_done=0,
    )

    db.add(task)
    db.commit()
    db.refresh(task)

    return task


@router.patch("/tasks/{task_id}/toggle", response_model=Task)
def toggle_task(task_id: int, db: Session = Depends(get_db)):
    task = db.query(TaskModel).filter(TaskModel.id == task_id).first()

    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")

    if task.is_done:
        task.is_done = 0
        task.completed_at = None
    else:
        task.is_done = 1
        task.completed_at = datetime.utcnow()

    db.commit()
    db.refresh(task)

    return task

@router.patch("/tasks/{task_id}", response_model=Task)
def update_task(
    task_id: int,
    task_data: TaskUpdate,
    db: Session = Depends(get_db),
):
    task = db.query(TaskModel).filter(TaskModel.id == task_id).first()

    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")

    if not task_data.title.strip():
        raise HTTPException(status_code=400, detail="Task title is required")

    task.title = task_data.title.strip()

    db.commit()
    db.refresh(task)

    return task


@router.delete("/tasks/{task_id}")
def delete_task(task_id: int, db: Session = Depends(get_db)):
    task = db.query(TaskModel).filter(TaskModel.id == task_id).first()

    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")

    db.delete(task)
    db.commit()

    return {
        "status": "ok",
        "message": "Task deleted successfully",
    }

@router.patch("/tasks/{task_id}/today", response_model=Task)
def toggle_task_today(task_id: int, db: Session = Depends(get_db)):
    task = db.query(TaskModel).filter(TaskModel.id == task_id).first()

    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")

    task.is_today = 0 if task.is_today else 1

    db.commit()
    db.refresh(task)

    return task

@router.patch("/tasks/{task_id}/notes", response_model=Task)
def update_task_notes(
    task_id: int,
    task_notes_data: TaskNotesUpdate,
    db: Session = Depends(get_db),
):
    task = db.query(TaskModel).filter(TaskModel.id == task_id).first()

    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")

    task.notes = (
        task_notes_data.notes.strip()
        if task_notes_data.notes
        else None
    )

    db.commit()
    db.refresh(task)

    return task

File: frontend/src/api/tasks.js
const API_BASE_URL = "http://localhost:8000";

async function parseError(response, fallbackMessage) {
  let errorMessage = fallbackMessage;

  try {
    const errorData = await response.json();
    errorMessage = errorData.detail || fallbackMessage;
  } catch {
    // Ignore parsing error
  }

  return errorMessage;
}

export async function getTaskLists() {
  const response = await fetch(`${API_BASE_URL}/task-lists`);

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to fetch task lists"));
  }

  return response.json();
}

export async function createTaskList(taskListData) {
  const response = await fetch(`${API_BASE_URL}/task-lists`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(taskListData),
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to create task list"));
  }

  return response.json();
}

export async function getTasks(taskListId) {
  const response = await fetch(`${API_BASE_URL}/task-lists/${taskListId}/tasks`);

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to fetch tasks"));
  }

  return response.json();
}

export async function createTask(taskListId, taskData) {
  const response = await fetch(`${API_BASE_URL}/task-lists/${taskListId}/tasks`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(taskData),
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to create task"));
  }

  return response.json();
}

export async function toggleTask(taskId) {
  const response = await fetch(`${API_BASE_URL}/tasks/${taskId}/toggle`, {
    method: "PATCH",
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to toggle task"));
  }

  return response.json();
}

export async function updateTask(taskId, taskData) {
  const response = await fetch(`${API_BASE_URL}/tasks/${taskId}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(taskData),
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to update task"));
  }

  return response.json();
}

export async function deleteTask(taskId) {
  const response = await fetch(`${API_BASE_URL}/tasks/${taskId}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to delete task"));
  }

  return response.json();
}

export async function updateTaskList(taskListId, taskListData) {
  const response = await fetch(`${API_BASE_URL}/task-lists/${taskListId}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(taskListData),
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to update task list"));
  }

  return response.json();
}

export async function deleteTaskList(taskListId) {
  const response = await fetch(`${API_BASE_URL}/task-lists/${taskListId}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to delete task list"));
  }

  return response.json();
}

export async function toggleTaskToday(taskId) {
  const response = await fetch(`${API_BASE_URL}/tasks/${taskId}/today`, {
    method: "PATCH",
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to update today status"));
  }

  return response.json();
}


export async function updateTaskNotes(taskId, taskNotesData) {
  const response = await fetch(`${API_BASE_URL}/tasks/${taskId}/notes`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(taskNotesData),
  });

  if (!response.ok) {
    throw new Error(await parseError(response, "Failed to update task notes"));
  }

  return response.json();
}